src/net/sourceforge/kolmafia/textui/javascript/AshStub.java

```diff
@@ -42,107 +35,24 @@ public String getFunctionName() {
-    // now every undefined value leads to an error anyway, so this is fine.
-    int definedArgs = args.length;
-    while (definedArgs > 0 && Undefined.isUndefined(args[definedArgs - 1])) {
-      definedArgs -= 1;
-    }
-
     ScriptableValueConverter coercer = new ScriptableValueConverter(cx, scope);
-
-    // Find library function matching arguments, in two stages.
-    // First, designate any arguments where we can't determine type (or aggregate type) from JS
-    // as ANY_TYPE, which will be replaced in findMatchingFunction with the target type
-    // to force a match. This is mainly relevant for empty arrays and records.
-    List<Value> ashArgs = new ArrayList<>();
-    for (int i = 0; i < definedArgs; i++) {
-      Object original = args[i];
-      if (Undefined.isUndefined(original)) {
-        throw new EvaluatorException("Passing undefined to an ASH function is not supported.");
-      }
-      if (original == null) {
-        throw new EvaluatorException("Passing null to an ASH function is not supported.");
-      }
-      try {
-        Value coerced = coercer.fromJava(original);
-        if (coerced == null
-            || (coerced.getType() instanceof AggregateType agg
-                && agg.getDataType().equals(DataTypes.ANY_TYPE))) {
-          coerced = new Value(DataTypes.ANY_TYPE);
-        }
-        ashArgs.add(coerced);
-      } catch (ValueConverter.ValueConverterException e) {
-        throw new EvaluatorException(e.getMessage());
-      }
-    }
-    Function function = findMatchingFunction(ashArgs, true);
-
-    if (function == null) {
-      throw controller.runtimeException(Parser.undefinedFunctionMessage(ashFunctionName, ashArgs));
-    }
-
-    // Second, infer the type for any ANY_TYPE arguments from the closest function match.
-    boolean argsChanged = false;
-    for (int i = 0; i < ashArgs.size(); i++) {
-      if (ashArgs.get(i).getType() != DataTypes.ANY_TYPE) {
-        continue;
-      }
-      Object original = args[i];
-      // Try again, this time with a type hint.
-      Type typeHint = function.getVariableReferences().get(i).getType();
-      Value coerced = coercer.fromJava(original, typeHint);
-      if (coerced == null || coerced.getType() == DataTypes.ANY_TYPE) {
-        throw controller.runtimeException("Could not coerce argument to valid ASH value.");
-      }
-      ashArgs.set(i, coerced);
-      argsChanged = true;
-    }
-    if (argsChanged) {
-      function = findMatchingFunction(ashArgs, false);
-
-      if (function == null) {
+    Value ashReturnValue;
+    try {
+      var functionWithArgs =
+          coercer.findMatchingFunctionConvertArgs(getAllFunctions(), ashFunctionName, args);
+      if (functionWithArgs == null || functionWithArgs.function() == null) {
         throw controller.runtimeException(
-            Parser.undefinedFunctionMessage(ashFunctionName, ashArgs));
+            Parser.undefinedFunctionMessage(ashFunctionName, List.of()));
```

Comment:
okay so still a problem here, I can make it crash:

```
> js var a = require("kolmafia"); a.monsterHp(1,{})

Unexpected error, debug log printed.

> js var a = require("kolmafia"); a.monsterHp(1,[])

Unexpected error, debug log printed.

> js var a = require("kolmafia"); a.monsterHp(1,/1/)

Unexpected error, debug log printed.
```

`null` and `undefined` both work which is good!

Test coverage for this would also be appreciated.

Reply:
```
> js var a = require("kolmafia"); a.monsterHp(1,/1/)
```
still prints a debug log.

While you catch a `ValueConverterException` from `coercer.fromJava(arg)`, I don't think that function ever throws the exception -- rather it just returns null. I think converting null to any is the right move here (or that will fix the debug log issue, at least).